backup-tools/backup-replication-image/go-src/job_queue.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	_ "expvar"
+	"log"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ncw/swift"
+	"github.com/sapcc/containers/backup-tools/go-src/configuration"
+	"github.com/sapcc/containers/backup-tools/go-src/swiftcli"
+	"github.com/sapcc/containers/backup-tools/go-src/utils"
+)
+
+const maxWorkers = 5 // normal 5; debug 2
+
+var (
+	alreadyPrinted   int
+	currentFilesDone = 0
+	expiration       int64
+	backupDir        = "/backup/tmp"
+	EnvFrom          *Env
+	EnvTo            = make([]*Env, 2)
+)
+
+type Env struct {
+	Cfg      configuration.EnvironmentStruct
+	Files    []string
+	SwiftCli *swift.Connection
+}
+
+type Job struct {
+	EnvFrom      *Env
+	EnvTo        []*Env
+	File         string
+	FileAllCount int
+	FileNumber   int
+}
+
+func doWork(id int, j Job) {
+	var dlErr, upErr error
+	var uploadDone bool
+	var dlFilePath string
```

Comment:
It's bad style to declare variables at the start of the function. Declare them when you use them for the first time, with the `:=` operator, and take advantage of the variables going out of scope at the end of their block.

Reply:
there is no alternative as to do it in this way. If there was a way to do so I would do it. Or what would you do there? each if is a own block and each for, switch so... where should I pre define it when not there for the loop and the if statements?

Reply:
This is what I can immediately refactor to declare variables more locally: https://gist.github.com/majewsky/7781e91d0921ced5cb5fe56cbc161f14

From there, you can refactor the function to get rid of the goto and the otherwise strange control flow. The logic would go something like this:

```go
if file replicated to all regions {
  return
}
path, err := download file
if err != nil {
  delete path
  complain
  return
}
foreach target region {
  upload file to region || complain
}
```